File: Tree/ValidBST.c
/*Given a binary tree, determine if it is a valid binary search tree (BST).

Assume a BST is defined as follows:

The left subtree of a node contains only nodes with keys less than the node’s key.
The right subtree of a node contains only nodes with keys greater than the node’s key.
Both the left and right subtrees must also be binary search trees.
Example :

Input : 
   1
  /  \
 2    3

Output : 0 or False


Input : 
  2
 / \
1   3

Output : 1 or True */


/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 * 
 * typedef struct TreeNode treenode;
 * 
 * treenode* treenode_new(int val) {
 *     treenode* node = (treenode *) malloc(sizeof(treenode));
 *     node->val = val;
 *     node->left = NULL;
 *     node->right = NULL;
 *     return node;
 * }
 */
/**
 * @input A : Root pointer of the tree 
 * 
 * @Output Integer
 * 
 */
//  #define INT1_MAX  4294967296
//  #define INT1_MIN -4294967296
 int isBST(treenode* root,int minvalue, int maxvalue){
     if(root==NULL) return 1;
     if(root->val > minvalue && root->val < maxvalue
        && isBST(root->left,minvalue,root->val)
        && isBST(root->right,root->val,maxvalue))
        return 1;
    else{
        return 0;
    }
 }
int isValidBST(treenode* A) {
    int result = isBST(A,INT_MIN,INT_MAX);
    return result;
}

File: Tree/ValidBST.py
__author__ = 'mushahidalam'
import sys


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    #Attempt 2
    inorder_output = list()
    def __init__(self):
        self.inorder_output = []

    def inorderTraversal(self, root):
        """
        :param root: TreeNode
        :return: None
        """
        if root == None:
            return
        self.inorderTraversal(root.left)
        self.inorder_output.append(root.val)
        self.inorderTraversal(root.right)

    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root == None:
            return True
        self.inorderTraversal(root)
        if len(self.inorder_output) == 1:
            return True
        for i in range(len(self.inorder_output) - 1):
            if self.inorder_output[i] < self.inorder_output[i+1]:
                continue
            return False
        return True

    #Previous Accepted Recursive Solution
    # def checkBST(self, root, min, max):
    #
    #     if root == None:
    #         return 1
    #     if root.val > min and root.val <= max:
    #         return 1
    #     return (self.checkBST(root.left, min, root.val) and self.checkBST(root.right, root.val, max))

    # def isValidBST(self, root):
    #     """
    #     :type root: TreeNode
    #     :rtype: bool
    #     """
    #     self.checkBST(root, -sys.maxint, sys.maxint)
obj = Solution()

root = TreeNode(2)
root.left = TreeNode(1)
root.right = TreeNode(3)
print obj.isValidBST(root)

root = TreeNode(5)
root.left = TreeNode(1)
root.right = TreeNode(4)
root.right.left = TreeNode(3)
root.right.right = TreeNode(6)
print obj.isValidBST(root)
